File: ruby/lib/steam/community/game_weapon.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2009-2011, Sebastian Staudt

# A module implementing basic functionality for classes representing game
# weapons
#
# @author Sebastian Staudt
module GameWeapon

  # Returns the number of kills achieved with this weapon
  #
  # @return [Fixnum] The number of kills achieved
  attr_reader :kills

  # Returns the unique identifier for this weapon
  #
  # @return [String] The identifier of this weapon
  attr_reader :id

  # Returns the number of shots fired with this weapon
  #
  # @return [Fixnum] The number of shots fired
  attr_reader :shots

  # Creates a new game weapon instance with the data provided
  #
  # @param [REXML::Element] weapon_data The data representing this weapon
  def initialize(weapon_data)
    @kills = weapon_data.elements['kills'].text.to_i
  end

  # Returns the average number of shots needed for a kill with this weapon
  #
  # @return [Float] The average number of shots needed for a kill
  def avg_shots_per_kill
    @shots.to_f / @kills
  end

end

File: php/lib/steam/community/GameWeapon.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2009, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage Steam Community
 */

/**
 * Super class for classes representing game weapons
 * @package Steam Condenser (PHP)
 * @subpackage Steam Community
 */
abstract class GameWeapon {

    private $avgShotsPerKill;

    protected $kills;

    protected $id;

    protected $shots;

    public function __construct($weaponData) {
        $this->kills = (int) $weaponData->kills;
    }

    /**
     * Returns the average number of shots needed for a kill with this weapon.
     * Calculates the value if needed.
     * @return float
     */
    public function getAvgShotsPerKill() {
        if(empty($this->avgShotsPerKill)) {
            $this->avgShotsPerKill = $this->shots / $this->kills;
        }

        return $this->avgShotsPerKill;
    }

    public function getId() {
        return $this->id;
    }

    public function getKills() {
        return $this->kills;
    }

    public function getShots() {
        return $this->shots;
    }

}
?>
